src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -373,44 +373,24 @@ absl::StatusOr<bool> PrivateKeyAndCertificateMatch(
   if (cert_chain.empty()) {
     return absl::InvalidArgumentError("Certificate string is empty.");
   }
-  BIO* cert_bio = BIO_new_mem_buf(cert_chain.data(), cert_chain.size());
-  if (cert_bio == nullptr) {
-    return absl::InvalidArgumentError(
-        "Conversion from certificate string to BIO failed.");
-  }
-  // Reads the first cert from the cert_chain which is expected to be the leaf
-  // cert
-  X509* x509 = PEM_read_bio_X509(cert_bio, nullptr, nullptr, nullptr);
-  BIO_free(cert_bio);
-  if (x509 == nullptr) {
+  OwnedOpenSslX509 x509(cert_chain.data(), cert_chain.size());
+  if (!x509.get_x509()) {
     return absl::InvalidArgumentError(
         "Conversion from PEM string to X509 failed.");
   }
-  EVP_PKEY* public_evp_pkey = X509_get_pubkey(x509);
-  X509_free(x509);
-  if (public_evp_pkey == nullptr) {
+  OwnedOpenSslPrivateKey public_evp_pkey(X509_get_pubkey(x509.get_x509()));
```

Comment:
Oh so this can also include public key. Let's change the class name to `OwnedOpenSslPKey` then.
Also I feel like there is no need to call it `Owned*`. Maybe just `OpenSslX509` and `OpenSslPKey` would be sufficient.